bootstrap/bukkit/src/main/java/org/geysermc/platform/bukkit/world/GeyserBukkitWorldManager.java

```diff
@@ -28,18 +28,43 @@
 
 import com.github.steveice10.mc.protocol.data.game.world.block.BlockState;
 
+import lombok.AllArgsConstructor;
 import org.bukkit.Bukkit;
+import org.bukkit.block.Block;
 import org.geysermc.connector.network.session.GeyserSession;
 import org.geysermc.connector.network.translators.world.WorldManager;
 import org.geysermc.connector.network.translators.world.block.BlockTranslator;
+import us.myles.ViaVersion.protocols.protocol1_13_1to1_13.Protocol1_13_1To1_13;
+import us.myles.ViaVersion.protocols.protocol1_15to1_14_4.data.MappingData;
 
+@AllArgsConstructor
 public class GeyserBukkitWorldManager extends WorldManager {
 
+    private final boolean isLegacy;
+    // You need ViaVersion to connect to an older server with Geyser.
+    // However, we still check for ViaVersion in case there's some other way that gets Geyser on a pre-1.13 Bukkit server
+    private final boolean isViaVersion;
+
     @Override
+    @SuppressWarnings("deprecation")
     public BlockState getBlockAt(GeyserSession session, int x, int y, int z) {
         if (session.getPlayerEntity() == null) {
             return BlockTranslator.AIR;
         }
+        if (isLegacy) {
+            if (isViaVersion) {
+                Block block = Bukkit.getPlayer(session.getPlayerEntity().getUsername()).getWorld().getBlockAt(x, y, z);
+                // Black magic that gets the old block state ID
+                int oldBlockId = (block.getType().getId() << 4) | (block.getData() & 0xF);
+                // Convert block state from old version -> 1.13 -> 1.13.1 -> 1.14 -> 1.15
+                int thirteenBlockId = us.myles.ViaVersion.protocols.protocol1_13to1_12_2.data.MappingData.blockMappings.getNewId(oldBlockId);
+                int thirteenPointOneBlockId = Protocol1_13_1To1_13.getNewBlockStateId(thirteenBlockId);
+                int fourteenBlockId = us.myles.ViaVersion.protocols.protocol1_14to1_13_2.data.MappingData.blockStateMappings.getNewId(thirteenPointOneBlockId);
+                return new BlockState(MappingData.blockStateMappings.getNewId(fourteenBlockId));
+            } else {
```

Comment:
This else should be moved to the bracket below since if the user is using a legacy version and does not have ViaVersion (for whatever reason), they’ll still get an exception of a missing method. Even though Geyser pretty much needs ViaVersion to run on these older versions, it’s good to have this moved down just incase.

Reply:
Read this wrong nvm